pkg/agent/flowexporter/connections/connections.go

```diff
@@ -28,42 +27,64 @@ import (
 var _ ConnectionStore = new(connectionStore)
 
 type ConnectionStore interface {
-	Run(stopCh <-chan struct{})
+	Run(stopCh <-chan struct{}, pollDone chan bool)
+	IterateCxnMapWithCB(updateCallback flowexporter.FlowRecordUpdate) error
+	FlushConnectionStore()
 }
 
 type connectionStore struct {
-	connections map[flowexporter.ConnectionKey]flowexporter.Connection // Add 5-tuple as string array
-	connDumper  ConnTrackDumper
-	ifaceStore  interfacestore.InterfaceStore
-	mutex       sync.Mutex
+	// pollDone channel is used for synchronization of poll(ConnectionStore.Run) and export(FlowExporter.Run) go routines.
+	// Therefore, there is no requirement of lock to make connections map thread safe.
+	connections    map[flowexporter.ConnectionKey]flowexporter.Connection // Add 5-tuple as string array
+	connDumper     ConnTrackDumper
+	ifaceStore     interfacestore.InterfaceStore
+	pollInterval   time.Duration
+	exportInterval time.Duration
 }
 
-func NewConnectionStore(ctDumper ConnTrackDumper, ifaceStore interfacestore.InterfaceStore) *connectionStore {
+func NewConnectionStore(ctDumper ConnTrackDumper, ifaceStore interfacestore.InterfaceStore, pollInterval time.Duration, exportInterval time.Duration) *connectionStore {
 	return &connectionStore{
-		connections: make(map[flowexporter.ConnectionKey]flowexporter.Connection),
-		connDumper:  ctDumper,
-		ifaceStore:  ifaceStore,
+		connections:    make(map[flowexporter.ConnectionKey]flowexporter.Connection),
+		connDumper:     ctDumper,
+		ifaceStore:     ifaceStore,
+		pollInterval:   pollInterval,
+		exportInterval: exportInterval,
 	}
 }
 
 // Run polls the connTrackDumper module periodically to get connections. These connections are used
 // to build connection store.
-func (cs *connectionStore) Run(stopCh <-chan struct{}) {
+func (cs *connectionStore) Run(stopCh <-chan struct{}, pollDone chan bool) {
 	klog.Infof("Starting conntrack polling")
 
-	ticker := time.NewTicker(flowexporter.PollInterval)
-	defer ticker.Stop()
+	pollTicker := time.NewTicker(cs.pollInterval)
+	defer pollTicker.Stop()
+	exportTrigger := uint(cs.exportInterval / cs.pollInterval)
+	exportCounter := uint(0)
+
 	for {
 		select {
 		case <-stopCh:
 			break
-		case <-ticker.C:
+		case <-pollTicker.C:
+			if exportCounter == 0 {
+				exportCounter = exportTrigger
+				// Flush connection map once all flow records are sent.
+				// TODO: Optimize this logic by flushing individual connections based on the individual timeout values.
```

Comment:
what are the consequences of flushing unconditionally? do we end up with some incorrect counters in the IPFIX flow records (for the `delta` fields)? if yes that's not good.

Reply:
Records keep separate copy of stats to store previous stats and not connected to the connection object. However, changed the deletion logic for connections in connection map using an isActive flag rather than doing unoptimized way of cleaning memory and adding the connection back again.